drift/lib/src/runtime/manager/references.dart

```diff
@@ -0,0 +1,492 @@
+
+  /// This callback is used to prefetch referenced data and insert it into the TypedResult object.
+  late final TypedResultTransformer withPrefetches;
+
+  /// Create a [PrefetchHooks] object
+  PrefetchHooks(
+      {StateTransformer? addJoins, TypedResultTransformer? withPrefetches}) {
+    withJoins = addJoins ?? _defaultStateTransformer;
+    this.withPrefetches = withPrefetches ?? (results) async => results;
+  }
+}
+
+/// This class is used to convert a table which expects a single dataclass per row into a table that can contain a multiple dataclasses per row.
+/// However when we prefetch multiple references, we need to store them in a list, which isn't supported by the referenced table class.
+/// For single references, we use original referenced table as a key in the TypedResult._parsedData object.
+/// {@macro manager_internal_use_only}
+class MultiTypedResultKey<$Table extends Table, $Dataclass>
+    implements ResultSetImplementation<$Table, $Dataclass> {
+  @override
+  final List<GeneratedColumn<Object>> $columns;
+
+  @override
+  final $Table asDslTable;
+
+  @override
+  final DatabaseConnectionUser attachedDatabase;
+
+  @override
+  final Map<String, GeneratedColumn<Object>> columnsByName;
+
+  @override
+  final String entityName;
+
+  @override
+  FutureOr<$Dataclass> map(Map<String, dynamic> data, {String? tablePrefix}) =>
+      _mapFunction(data, tablePrefix: tablePrefix);
+
+  final FutureOr<$Dataclass> Function(Map<String, dynamic> data,
+      {String? tablePrefix}) _mapFunction;
+
+  @override
+  String get aliasedName => entityName;
+
+  @override
+  MultiTypedResultKey<$Table, $Dataclass> createAlias(String alias) {
+    return MultiTypedResultKey._(
+      $columns: $columns,
+      asDslTable: asDslTable,
+      attachedDatabase: attachedDatabase,
+      columnsByName: columnsByName,
+      entityName: alias,
+      mapFunction: _mapFunction,
+    );
+  }
+
+  const MultiTypedResultKey._(
+      {required this.$columns,
+      required this.asDslTable,
+      required this.attachedDatabase,
+      required this.columnsByName,
+      required this.entityName,
+      required FutureOr<$Dataclass> Function(Map<String, dynamic>,
+              {String? tablePrefix})
+          mapFunction})
+      : _mapFunction = mapFunction;
+
+  /// Create a [MultiTypedResultKey] from a table
+  static MultiTypedResultKey<$Table, List<$Dataclass>>
+      fromTable<$Table extends Table, $Dataclass>(
+          TableInfo<$Table, $Dataclass> table,
+          {required String aliasName}) {
+    return MultiTypedResultKey<$Table, List<$Dataclass>>._(
+      $columns: table.$columns,
+      asDslTable: table.asDslTable,
+      attachedDatabase: table.attachedDatabase,
+      columnsByName: table.columnsByName,
+      entityName: aliasName,
+
+      /// This table is used a key for a map and should never be used to read data
+      /// However, in case it is, we will map a single item to a list of items
```

Comment:
Supporting things we don't want to support doesn't seem right to me. Either we officially support this, or we override the relevant methods to throw an `UnsupportedError`.